WordPress/Classes/Utility/WebViewAuthenticator.swift

```diff
@@ -0,0 +1,206 @@
+        cookieJar.hasCookie(url: loginURL, username: username) { [weak self] (hasCookie) in
+            guard let authenticator = self else {
+                return
+            }
+
+            let request = authenticator.request(url: url, authenticated: !hasCookie)
+            completion(request)
+        }
+    }
+
+    /// Intercepts and rewrites any potential redirect after login.
+    ///
+    /// This should be called whenever a web view needs to decide if it should
+    /// load a request. If this returns a non-nil value, the resulting request
+    /// should be loaded instead.
+    ///
+    /// - Parameters:
+    ///     - request: the request that was going to be loaded by the web view.
+    ///
+    /// - Returns: a request to be loaded instead. If `nil`, the original
+    /// request should continue loading.
+    ///
+    func interceptRedirect(request: URLRequest) -> URLRequest? {
+        guard let url = request.url,
+            let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
+            components.scheme == "https",
+            components.host == "wordpress.com",
+            let encodedRedirect = components
+                .queryItems?
+                .first(where: { $0.name == WebViewAuthenticator.redirectParameter })?
+                .value,
+            let redirect = encodedRedirect.removingPercentEncoding,
+            let redirectUrl = URL(string: redirect)
+            else {
+                return nil
+        }
+
+        return URLRequest(url: redirectUrl)
+    }
+}
+
+private extension WebViewAuthenticator {
+    func request(url: URL, authenticated: Bool) -> URLRequest {
+        guard authenticated else {
+            return unauthenticatedRequest(url: url)
+        }
+        return authenticatedRequest(url: url) ?? unauthenticatedRequest(url: url)
+    }
+
+    func authenticatedRequest(url: URL) -> URLRequest? {
+        var request = URLRequest(url: loginURL)
+        request.httpMethod = "POST"
+        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+        request.httpBody = body(url: url)
+        if let authToken = authToken {
+            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
+        }
+        return request
+    }
+
+    func unauthenticatedRequest(url: URL) -> URLRequest {
+        return URLRequest(url: url)
+    }
+
+    func body(url: URL) -> Data? {
+        guard let encodedUsername = username.urlFormEncoded,
+            let encodedUrl = redirectUrl(url: url.absoluteString)?.urlFormEncoded else {
+            return nil
+        }
+        let encodedPassword = password?.urlFormEncoded
+
+        var parameters = "log=\(encodedUsername)"
+        if let encodedPassword = encodedPassword {
+            parameters += "&pwd=\(encodedPassword)"
+        }
+        parameters += "&redirect_to=\(encodedUrl)"
+
```

Comment:
I don't really like composing query strings manually, what do you think about using `URLComponents` ?

```
    var queryItems = [URLQueryItem]()
    queryItems.append(URLQueryItem(name: "log", value: encodedUsername))
    if let encodedPassword = password?.urlFormEncoded {
         queryItems.append(URLQueryItem(name: "pwd", value: encodedPassword))
    }
    queryItems.append(URLQueryItem(name: "redirect_to", value: encodedUrl))
    var components = URLComponents()
    components.queryItems = queryItems

    return components.query?.data(using: .utf8)
```

Reply:
> what do you think about using URLComponents ?

I think that I can't believe I missed that when that's what I used for decoding the wpios_redirect parameter 😮 Thanks for the suggestion.